### PR TITLE
Add MultiModallity PerceiverIO Implementation

### PR DESCRIPTION
**Pull request by jacobbieker**
Multimodality PerceiverIO wrapper essentially. Taken from Satflow here https://github.com/openclimatefix/satflow/blob/main/satflow/models/perceiver.py 

Starting point for a general/combined PV/SatFlow Perceiver implementation?

### CODE REVIEW

File: perceiver_pytorch/encoders.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
+import torch
+import torch.nn.functional as F
+import numpy as np
+import math
+import einops
+
+def extract_image_patches(x, kernel, stride=1, dilation=1):
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:7]: worth putting this function, in utils.py file as maybe it doesnt fi under enconders?

Comment (peterdudfield) [reply perceiver_pytorch/encoders.py:7]: Doc string would be great for what i does, , and typing on input variables is very useful

File: perceiver_pytorch/encoders.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    # Do TF 'SAME' Padding
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:8]: what does this mean?

File: perceiver_pytorch/encoders.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    b, c, h, w = x.shape
+    h2 = math.ceil(h / stride)
+    w2 = math.ceil(w / stride)
+    pad_row = (h2 - 1) * stride + (kernel - 1) * dilation + 1 - h
+    pad_col = (w2 - 1) * stride + (kernel - 1) * dilation + 1 - w
+    x = F.pad(x, (pad_row // 2, pad_row - pad_row // 2, pad_col // 2, pad_col - pad_col // 2))
+
+    # Extract patches
+    patches = x.unfold(2, kernel, stride).unfold(3, kernel, stride)
+    patches = patches.permute(0, 4, 5, 1, 2, 3).contiguous()
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:18]: I dont understand this bit

File: perceiver_pytorch/encoders.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+    return patches.view(b, -1, patches.shape[-2], patches.shape[-1])
+
+
+def reverse_space_to_depth(
+        frames: np.ndarray, temporal_block_size: int = 1, spatial_block_size: int = 1
+) -> np.ndarray:
+    """Reverse space to depth transform."""
+    if len(frames.shape) == 4:
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:27]: Could just add a comment, saying dim=4 is an image, dim=5 is a video

File: perceiver_pytorch/encoders.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            dw=spatial_block_size,
+        )
+    else:
+        raise ValueError(
+            "Frames should be of rank 4 (batch, height, width, channels)"
+            " or rank 5 (batch, time, height, width, channels)"
+        )
+
+
+class Conv2DDownsample(torch.nn.Sequential):
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:75]: Should these functions be a separate file called 'convolution'?

File: perceiver_pytorch/encoders.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+            use_batchnorm: bool = True,
+    ):
+        """
+        Constructs a Conv2DDownsample model
+        Args:
+            num_layers: Number of conv -> maxpool layers
+            output_channels: Number of output channels
+            input_channels: Number of input channels to first layer
+            use_batchnorm: Whether to use Batch Norm
+        """
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:90]: nice

File: perceiver_pytorch/encoders.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    elif len(frames.shape) == 5:
+        return einops.rearrange(
+            frames,
+            "b (t dt) (h dh) (w dw) c -> b t h w (dt dh dw c)",
+            dt=temporal_block_size,
+            dh=spatial_block_size,
+            dw=spatial_block_size,
+        )
+    else:
+        raise ValueError(
```
Comment (peterdudfield) [inline perceiver_pytorch/encoders.py:69]: very nice

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
+from perceiver_pytorch.perceiver_pytorch import fourier_encode
+from perceiver_pytorch.perceiver_io import PerceiverIO
+import torch
+from typing import List, Iterable, Dict, Optional, Any
+from einops import rearrange, repeat
+from dataclasses import dataclass
+
+
+@dataclass
```
Comment (peterdudfield) [inline perceiver_pytorch/multi_perceiver_pytorch.py:9]: I like it

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+class InputModality:
+    name: str
+    input_channels: int
+    input_axis: int
+    num_freq_bands: int
+    max_freq: float
+    freq_base: int = 2
+    sin_only: bool = False
+    fourier_encode: bool = True
+
```
Comment (peterdudfield) [inline perceiver_pytorch/multi_perceiver_pytorch.py:19]: Worth explaining in comments what they all are?

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        return input_dim
+
+
+def modality_encoding(batch_size: int, axes, modality_index: int, num_modalities: int) -> torch.Tensor:
+    """
+    Return one-hot encoding of modality given num_modalities, batch size and axes.
+    The result need to be compatible with the modality data for concatenation.
+    :param modality_index:
+    :param num_modalities:
+    :return:
```
Comment (peterdudfield) [inline perceiver_pytorch/multi_perceiver_pytorch.py:35]: fill in?

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        self.output_channels = output_channels
+        self.modalities = {modality.name: modality for modality in modalities}
+        # we encode modality with one hot encoding, so need one dim per modality:
+        modality_encoding_dim = sum([1 for _ in modalities])
+        # input_dim is the maximum dimension over all input modalities:
+        input_dim = max(modality.input_dim for modality in modalities) + modality_encoding_dim
+        # Pop dim
+        self.max_modality_dim = input_dim
+        kwargs.pop("dim")
+        # Want toe logit_dim to be the same as the channels * width or height
```
Comment (peterdudfield) [inline perceiver_pytorch/multi_perceiver_pytorch.py:84]: good comments

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            linearized_data.append(data)
+
+        # Concatenate all the modalities:
+        data = torch.cat(linearized_data, dim=1)
+
+        # After this is the PerceiverIO backbone, still would need to decode it back to an image though
+        # Should include the query shape here for the output we want, could be learned embeddings, repeated input frames of the same shape that is desired, etc.
+        perceiver_output = self.perceiver.forward(data, mask, queries)
+
+        # For multiple modalities, they are split after this beack into different tensors
```
Comment (peterdudfield) [inline perceiver_pytorch/multi_perceiver_pytorch.py:152]: typoe

File: perceiver_pytorch/encoders.py (hunk 5 above)
Comment (jacobbieker) [reply perceiver_pytorch/encoders.py]: Done!

File: perceiver_pytorch/encoders.py (hunk 1 above)
Comment (jacobbieker) [reply perceiver_pytorch/encoders.py]: I've split up the code into smaller files, and am working on adding typing now!

File: perceiver_pytorch/encoders.py (hunk 4 above)
Comment (jacobbieker) [reply perceiver_pytorch/encoders.py]: Done

File: perceiver_pytorch/encoders.py (hunk 3 above)
Comment (jacobbieker) [reply perceiver_pytorch/encoders.py]: Added comments from the original source

File: perceiver_pytorch/multi_perceiver_pytorch.py (hunk 9 above)
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Added!

File: perceiver_pytorch/multi_perceiver_pytorch.py (hunk 10 above)
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Done!

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            output_channels: int = 12,
+            forecast_steps: int = 48,
+            sin_only: bool = False,
+            output_shape: int = 32,
+            **kwargs,
+    ):
+        """
+        PerceiverIO made to work more specifically with timeseries images
+        Not a recurrent model, so like MetNet somewhat, can optionally give a one-hot encoded vector for the future
+        timestep
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Done!

File: perceiver_pytorch/multi_perceiver_pytorch.py (hunk 12 above)
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Fixed!

**Comment by jacobbieker** [comment]
It now has tests, and everything passes and seems to work other than the ```Conv3DUpsample``` which gives the wrong sized output when upsampling.

File: tests/test_decoders.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    decoder = ImageDecoder(postprocess_type="conv1x1", output_channels=12, input_channels=48, spatial_upsample=4)
+    inputs = torch.randn(2, 3, 48, 64, 64)
+    with torch.no_grad():
+        out = decoder(inputs)
+    assert not torch.isnan(out).any(), "Output included NaNs"
+    # Conv1x1 downsample if spatial_upsample > 1
+    assert out.size() == (2, 3, 12, 16, 16)
+
+
+#def test_conv3d_video_decoder():
```
Comment (peterdudfield) [inline tests/test_decoders.py:64]: I like putting @pytest.mark.skip("This test doesn't work due too....  {name} {date} - put this as issue {link to issue}")

Might also be worth putting a comment in the function, say what is not working

Comment (peterdudfield) [reply tests/test_decoders.py:64]: Ah i see that is already commented in the function

Comment (jacobbieker) [reply tests/test_decoders.py]: Good idea! I've added that

File: perceiver_pytorch/utils.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    """
+    Extract image patches in a way similar to TensorFlow extract_image_patches
+    Taken from https://discuss.pytorch.org/t/tf-extract-image-patches-in-pytorch/43837/8
+    :param x: Input Torch Tensor
+    :param kernel: Size of kernel
+    :param stride: Stride of patch
+    :param dilation: Dilation rate
+    :return: Tensor of size [Batch, Height, Width, Channels*kernel*stride]
+    """
+    # Do TF 'SAME' Padding
```
Comment (JackKelly) [inline perceiver_pytorch/utils.py:20]: Might be worth copy-and-pasting @jacobbieker's GitHub comment into a code comment here: "In the Perceiver JAX implementation they extract image patches matching TensorFlow's SAME padding. PyTorch doesn't have that same kind of option, so this is a way to do that."

Comment (jacobbieker) [reply perceiver_pytorch/utils.py:24]: Done!

**Review by JackKelly — COMMENTED** [review]
Awesome work!  So cool to see this code split out as a general purpose module!  Thank you for all your work on this, @jacobbieker and @peterdudfield (and awesome review suggestions, @peterdudfield !)

One very general suggestion on variable naming:  I'd be keen to follow PyTorch's convention of prepending `num_` to the variable name whenever the variable counts the size of something.  For example, a variable named `layers` is ambiguous:  Is it the _number_ of layers, or a _list_ of layers?  Changing the variable name to `num_layers` makes it really explicit :)

File: perceiver_pytorch/convolutions.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+            channels = output_channels * pow(2, num_upsamples - 1 - i)
+            conv = torch.nn.ConvTranspose3d(
+                in_channels=input_channels,
+                out_channels=channels,
+                stride=(temporal_stride, space_stride, space_stride),
+                kernel_size=(4, 4, 4),
+            )
+            self.layers.append(conv)
+            if i != num_upsamples - i:
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/convolutions.py:116]: Yes, it is

File: perceiver_pytorch/modalities.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+@dataclass
+class InputModality:
+    name: str
+    input_channels: int
+    input_axis: int
+    num_freq_bands: int
+    max_freq: float
+    freq_base: int = 2
+    sin_only: bool = False
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/modalities.py:13]: Ah, I changed some and missed this one! Going with `sine_only`

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+class MultiPerceiver(torch.nn.Module):
+    def __init__(
+        self,
+        modalities: Iterable[InputModality],
+        fourier_encode_data: bool = True,
+        input_channels: int = 3,
+        output_channels: int = 12,
+        forecast_steps: int = 48,
+        sine_only: bool = False,
+        output_shape: int = 32,
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py:18]: I missed a bit copying and pasting between SatFlow and this one, this is now updated for being able to use tuples

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        modality_encoding_dim = sum([1 for _ in modalities])
+        # input_dim is the maximum dimension over all input modalities:
+        input_dim = (
+            max(modality.input_dim for modality in modalities) + modality_encoding_dim
+        )
+        # Pop dim
+        self.max_modality_dim = input_dim
+        kwargs.pop("dim", None)
+        # Want toe logit_dim to be the same as the channels * width or height
+        kwargs["logits_dim"] = output_shape * self.output_channels
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py:50]: Yes, you are right, I forgot a part from Satflow where I get the product of the tuple and use that to set the dim, readded it!

File: perceiver_pytorch/multi_perceiver_pytorch.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+            sorted(multi_modality_data.keys())
+        ):
+            assert (
+                modality_name in self.modalities
+            ), f"modality {modality_name} was not defined in constructor"
+            data = multi_modality_data[modality_name]
+            modality = self.modalities[modality_name]
+            b, *axis, _ = data.size()
+            assert len(axis) == modality.input_axis, (
+                f"input data must have the right number of  for modality {modality_name}. "
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Done!

File: perceiver_pytorch/experimental.py [hunk 21]
```diff
@@ -1,232 +1,234 @@
...
-                data = input_attn(data, mask=mask) + data
-                data = rev_cross_attn(data, context=x) + data
-                data = rev_cross_ff(data) + data
-
-            x = latent_attn(x) + x
-            x = latent_ff(x) + x
-
-        x = x.mean(dim=-2)
-        return self.to_logits(x)
+# Commented out by PD - 2021-09-02
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/experimental.py:1]: I'm not at all

File: perceiver_pytorch/decoders.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
+import torch
+import numpy as np
+from perceiver_pytorch.utils import reverse_space_to_depth
+from perceiver_pytorch.convolutions import Conv2DUpsample, Conv3DUpsample
+
+
+class ImageDecoder(torch.nn.Module):
+    def __init__(
+        self,
+        postprocess_type: str = "pixels",
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/decoders.py:10]: It probably makes sense to do that, each of the methods is entirely separate from the other ones. The benefit this has is swapping them out is as easy as changing the string passed to `prep_type` or `postprocessing_type` but we could make a similar wrapper for that if we wanted to.

[the remark this answers is not in the document]
Comment (JackKelly) [reply perceiver_pytorch/decoders.py:10]: Yeah, I really don't mind :).  It's certainly not urgent!

I'd expect separate classes (plus a wrapper to switch the class based on `prep_type`) to perhaps be a tiny bit easier to read & test; but I could be wrong (I haven't really thought precisely what separate classes would look like!)

File: perceiver_pytorch/decoders.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+        spatial_upsample: int = 1,
+        temporal_upsample: int = 1,
+        output_channels: int = -1,
+        input_channels: int = 12,
+        input_reshape_size=None,
+    ):
+        """
+        ImageDecoder modeled after JAX version here
+        https://github.com/deepmind/deepmind-research/blob/769bfdbeafbcb472cb8e2c6cfa746b53ac82efc2/perceiver/io_processors.py#L441-L510
+        :param postprocess_type: Type of postprocessing, one of conv, patches, pixels, raft, or conv1x1
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/decoders.py]: My preference would be Google's as well, I started changing the ones in this PR to it now too.

File: perceiver_pytorch/decoders.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+        :param spatial_upsample: How much to spatial upsample
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/decoders.py]: Added it.

File: perceiver_pytorch/multi_perceiver_pytorch.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (peterdudfield) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Yea Google is good

I'm a big fan of 'blacks' too
https://pypi.org/project/black/

[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/multi_perceiver_pytorch.py]: Yeah, I have it as a pre-commit hook in SatFlow to keep it consistent, might want to add it here too? Possibly in a smaller followup PR though haha

File: perceiver_pytorch/convolutions.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+        space_stride = 2
+        num_upsamples = max(num_space_upsamples, num_temporal_upsamples)
+        self.layers = torch.nn.ModuleList()
+        for i in range(num_upsamples):
+            if i >= num_temporal_upsamples:
+                temporal_stride = 1
+            if i >= num_space_upsamples:
+                space_stride = 1
+
+            channels = output_channels * pow(2, num_upsamples - 1 - i)
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/convolutions.py]: Done

File: perceiver_pytorch/convolutions.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+            conv = torch.nn.ConvTranspose3d(
+                in_channels=input_channels,
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/convolutions.py]: No, you are right, as the output channels of the previous layer would then be the input to the next

File: perceiver_pytorch/decoders.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jacobbieker) [reply perceiver_pytorch/decoders.py:10]: Okay! Yeah, I think it'd be easier to test and read as well. I'd probably go with changing it in a later PR? Possibly a bit more focused on the best way to make these decoder/encoders

[the remark this answers is not in the document]
Comment (JackKelly) [reply perceiver_pytorch/decoders.py:10]: Sure, yeah, no rush at all!  Another PR sounds like a good idea!

**Review by JackKelly — APPROVED** [review]
LGTM!  :)

**Pull request merged**